Std::length_error in running the camera_node

I installed the camera_ros from source. On my attempts to run the node using ros2 jazzy,  I am using a YUYV pixel format with a pi camera 3. Every resolution I try ends up with the node terminating after throwing an instance of 'std::length_error'. The error description saws that it cannot create std::vector larger than max_size().

Using #15 worked!

> this was built inside a container with the command:
> `docker run -it --network=host -v /dev:/dev/ -v /run/udev/:/run/udev/ --privileged --rm ros:humble bash`
> using the [build instructions](https://github.com/christianrauch/camera_ros?tab=readme-ov-file#source) 
>> I removed `--skip-keys=libcamera`,
   didn't compile/install any libcamera dependencies or libcamera itself
   and ran `git branch fix_dynamic_extent`
```bash
mkdir -p ~/camera_ws/
cd ~/camera_ws/
git clone https://github.com/christianrauch/camera_ros.git src/camera_ros
cd src/camera_ros
git checkout fix_dynamic_extent
cd ../..

apt update
source /opt/ros/humble/setup.bash
rosdep install --from-paths src --ignore-src -y #-y for automatic 'yes'
colcon build

source install/setup.bash
ros2 run camera_ros camera_node
```

---

> setting no parameters used the *NV21* format, resulting in `Unrecognized image encoding [nv21]`
> this is probably because ROS doesn't support that encoding?
> Anyway, I also set some other parameters to remove warnings
```console
root@bento-dev:~/camera_ws# ros2 run camera_ros camera_node --ros-args -p format:=XRGB8888 -p width:=1920 -p height:=1080 -p camera:=0
[0:24:41.552822524] [5042]  INFO Camera camera_manager.cpp:284 libcamera v0.1.0
[0:24:41.660784514] [5052]  INFO RPI vc4.cpp:390 Registered camera /base/soc/i2c0mux/i2c@0/imx708@1a to Unicam device /dev/media0 and ISP device /dev/media1
[WARN] [1718294411.663979311] [camera]: stream configuration adjusted from "1920x1080-XRGB8888" to "1920x1080-XRGB8888"
[0:24:41.662344396] [5042]  INFO Camera camera.cpp:1033 configuring streams: (0) 1920x1080-XRGB8888
[0:24:41.662978671] [5052]  INFO RPI vc4.cpp:512 Sensor: /base/soc/i2c0mux/i2c@0/imx708@1a - Selected sensor format: 2304x1296-SBGGR10_1X10 - Selected unicam format: 2304x1296-pBAA
[INFO] [1718294411.666719355] [camera]: camera "/base/soc/i2c0mux/i2c@0/imx708@1a" configured with 1920x1080-XRGB8888 stream
[WARN] [1718294411.669851100] [camera]: control NoiseReductionMode (39) not handled
[0:24:41.692182414] [5059]  WARN IPARPI ipa_base.cpp:1050 Could not set AF_PAUSE - no AF algorithm or not Continuous
[0:24:41.692278487] [5059]  WARN IPARPI ipa_base.cpp:1067 Could not set AF_TRIGGER - no AF algorithm or not Auto
[INFO] [1718294412.140485575] [camera]: using default calibration URL
[INFO] [1718294412.140828647] [camera]: camera calibration URL: file:///root/.ros/camera_info/imx708__base_soc_i2c0mux_i2c_0_imx708_1a_1920x1080.yaml
[ERROR] [1718294412.141129961] [camera_calibration_parsers]: Unable to open camera calibration file [/root/.ros/camera_info/imx708__base_soc_i2c0mux_i2c_0_imx708_1a_1920x1080.yaml]
[WARN] [1718294412.141239905] [camera]: Camera calibration file /root/.ros/camera_info/imx708__base_soc_i2c0mux_i2c_0_imx708_1a_1920x1080.yaml not found
[0:24:42.213161149] [5059]  WARN IPARPI ipa_base.cpp:1050 Could not set AF_PAUSE - no AF algorithm or not Continuous
[0:24:42.213275537] [5059]  WARN IPARPI ipa_base.cpp:1067 Could not set AF_TRIGGER - no AF algorithm or not Auto
```

---

Thank you for maintaining this package @christianrauch!
It's the best way to interface Pi cameras with ROS!

> Using #15 worked!

Nice. In the other issue it was reported that it causes another issue. Therefore I did not merge it yet.

Could you provide some details on how the `AfWindows` (the one that causes the issue) plays out in the end? Since I don't own the hardware, I cannot test how the `AfWindows` is exactly used. The `AfWindows` is a list of `Rectangle`, that as far as I understand, defines auto-focus areas in form of a bounding box made out of 4 values per "area".

If `AfWindows` only contains one `Rectangle`, then you should be able to get and set the parameter via a list of integers. If there are multiple `Rectangle` in `AfWindows`, then we cannot represent this in form of ROS parameters and the node should show a warning.

Could you please check the following:
1. start the node on the Raspberry Pi
2. check if the node prints `[ERROR] [<timestamp>] [camera]: unsupported control 'AfWindows' (type: Rectangle): ParameterValue only supported for arithmetic types`
3. get the parameter description and value via `ros2 param describe /camera AfWindows` and `ros2 param describe /camera AfWindows` (where `/camera` is the node name)


> setting no parameters used the _NV21_ format, resulting in `Unrecognized image encoding [nv21]` once a subscriber reads images
> this is probably because ROS doesn't support that encoding?

Correct. The encoding exists for the message, but a lot of tools in the ROS ecosystem do not support that encoding.

Could you provide a full log of the terminal output when running the node with the default parameters? Since https://github.com/christianrauch/camera_ros/pull/47, the node should select defaults that match the libcamera defaults when supported by the ROS message. Theoretically, the selected pixel format should be the same as the one selected by the `cam` example.

> Thank you for maintaining this package @christianrauch! It's the best way to interface Pi cameras with ROS!

Thanks :-)

> Edit: I just realized libcamera also works on my desktop PC. Well, i guess it's most modern camera software then, even better!

`libcamera` is quite versatile and aims to support all the "standard webcam" hardware. So I think this is a good replacement for the V4L2 and Raspberry Pi specific libraries.

**1&2:** The log in my last comment is the entire log, nothing else was written

**3:** `AfWindows` seems to contain one rectangle and makes sense
```console
root@frameslap:/bento_ws# ros2 param get /camera AfWindows 
Integer values are: array('q', [0, 0, 0, 0])

root@frameslap:/bento_ws# ros2 param describe /camera AfWindows
Parameter name: AfWindows
  Type: integer array
  Description: Rectangle array[] range {(0, 0)/0x0}..{(65535, 65535)/65535x65535} (default: {(0, 0)/0x0})
  Constraints:
```
> I fiddled around with the autofocus, everything seems to work fine
> only `AfTrigger` acts weird, as you need to toggle it for it to work
> (send a 1, nothing happens, then send a 0, and it triggers autofocus)

---

> Correct. The encoding exists for the message, but a lot of tools in the ROS ecosystem do not support that encoding.  

Thought so :+1:

---
The *entire* default log, for good measure:
```console
root@bento-dev:~/camera_ws# ros2 run camera_ros camera_node
[1:53:37.700525972] [3994]  INFO Camera camera_manager.cpp:284 libcamera v0.1.0
[1:53:37.854544722] [4004]  INFO RPI vc4.cpp:390 Registered camera /base/soc/i2c0mux/i2c@0/imx708@1a to Unicam device /dev/media1 and ISP device /dev/media3
[1:53:37.855800148] [4004]  WARN CameraSensorProperties camera_sensor_properties.cpp:249 No static properties available for 'imx708_wide'
[1:53:37.855871813] [4004]  WARN CameraSensorProperties camera_sensor_properties.cpp:251 Please consider updating the camera sensor properties database
[1:53:37.864540517] [4004]  INFO RPI vc4.cpp:390 Registered camera /base/soc/i2c0mux/i2c@1/imx708@1a to Unicam device /dev/media2 and ISP device /dev/media5
[INFO] [1718323520.386056317] [camera]: 
>> cameras:
   0: imx708 (/base/soc/i2c0mux/i2c@0/imx708@1a)
   1: imx708_wide (/base/soc/i2c0mux/i2c@1/imx708@1a)
[WARN] [1718323520.386842805] [camera]: no camera selected, using default: "/base/soc/i2c0mux/i2c@0/imx708@1a"
[WARN] [1718323520.386986044] [camera]: set parameter 'camera' to silent this warning
[INFO] [1718323520.387802569] [camera]: 
>> stream formats:
   - Pixelformat: NV21 (64x64 - 4608x2592)
   - Pixelformat: XBGR8888 (64x64 - 4608x2592)
   - Pixelformat: BGR888 (64x64 - 4608x2592)
   - Pixelformat: RGB888 (64x64 - 4608x2592)
   - Pixelformat: XRGB8888 (64x64 - 4608x2592)
   - Pixelformat: YUYV (64x64 - 4608x2592)
   - Pixelformat: UYVY (64x64 - 4608x2592)
[WARN] [1718323520.388042695] [camera]: no pixel format selected, using default: "NV21"
[WARN] [1718323520.388119768] [camera]: set parameter 'format' to silent this warning
[INFO] [1718323520.388332783] [camera]: 
>> NV21 format sizes:
   - 160x120
   - 240x160
   - 320x240
   - 400x240
   - 480x320
   - 640x360
   - 640x480
   - 720x480
   - 768x480
   - 854x480
   - 720x576
   - 800x600
   - 960x540
   - 1024x576
   - 960x640
   - 1024x600
   - 1024x768
   - 1280x720
   - 1152x864
   - 1280x800
   - 1360x768
   - 1366x768
   - 1440x900
   - 1280x1024
   - 1536x864
   - 1280x1080
   - 1600x900
   - 1400x1050
   - 1680x1050
   - 1600x1200
   - 1920x1080
   - 2048x1080
   - 1920x1200
   - 2160x1080
   - 2048x1152
   - 2560x1080
   - 2048x1536
   - 2560x1440
   - 2560x1600
   - 3840x1080
   - 2960x1440
   - 3440x1440
   - 2560x2048
   - 3200x1800
   - 3840x1600
   - 3200x2048
   - 3200x2400
   - 3840x2160
   - 4096x2160
   - 3840x2400
[WARN] [1718323520.388697778] [camera]: no dimensions selected, auto-selecting: "1920x1080"
[WARN] [1718323520.388751481] [camera]: set parameter 'width' or 'height' to silent this warning
[1:53:37.867970466] [3994]  INFO Camera camera.cpp:1033 configuring streams: (0) 1920x1080-NV21
[1:53:37.869038894] [4004]  INFO RPI vc4.cpp:512 Sensor: /base/soc/i2c0mux/i2c@0/imx708@1a - Selected sensor format: 2304x1296-SBGGR10_1X10 - Selected unicam format: 2304x1296-pBAA
[INFO] [1718323520.391931563] [camera]: camera "/base/soc/i2c0mux/i2c@0/imx708@1a" configured with 1920x1080-NV21 stream
[WARN] [1718323520.394777668] [camera]: control NoiseReductionMode (39) not handled
[1:53:37.888444678] [4013]  WARN IPARPI ipa_base.cpp:1050 Could not set AF_PAUSE - no AF algorithm or not Continuous
[1:53:37.888534658] [4013]  WARN IPARPI ipa_base.cpp:1067 Could not set AF_TRIGGER - no AF algorithm or not Auto
[INFO] [1718323520.831088415] [camera]: using default calibration URL
[INFO] [1718323520.831297931] [camera]: camera calibration URL: file:///root/.ros/camera_info/imx708__base_soc_i2c0mux_i2c_0_imx708_1a_1920x1080.yaml
[ERROR] [1718323520.831643648] [camera_calibration_parsers]: Unable to open camera calibration file [/root/.ros/camera_info/imx708__base_soc_i2c0mux_i2c_0_imx708_1a_1920x1080.yaml]
[WARN] [1718323520.831859108] [camera]: Camera calibration file /root/.ros/camera_info/imx708__base_soc_i2c0mux_i2c_0_imx708_1a_1920x1080.yaml not found
[1:53:38.420992929] [4013]  WARN IPARPI ipa_base.cpp:1050 Could not set AF_PAUSE - no AF algorithm or not Continuous
[1:53:38.421077872] [4013]  WARN IPARPI ipa_base.cpp:1067 Could not set AF_TRIGGER - no AF algorithm or not Auto
```
Note the Format, `1920x1080-NV21`
Interestingly the cam demo takes a different one per default `800x600-XRGB8888`:
```console
bento@bento-dev:~ $ cam -c1 --capture=1
[1:57:27.523006513] [21553]  INFO Camera camera_manager.cpp:313 libcamera v0.3.0+49-e76bb1f5
[1:57:27.809698111] [21554]  WARN RPiSdn sdn.cpp:40 Using legacy SDN tuning - please consider moving SDN inside rpi.denoise
[1:57:27.821443546] [21554]  INFO RPI vc4.cpp:400 Registered camera /base/soc/i2c0mux/i2c@0/imx708@1a to Unicam device /dev/media1 and ISP device /dev/media3
[1:57:27.825800925] [21554]  WARN CameraSensorProperties camera_sensor_properties.cpp:274 No static properties available for 'imx708_wide'
[1:57:27.825903979] [21554]  WARN CameraSensorProperties camera_sensor_properties.cpp:276 Please consider updating the camera sensor properties database
[1:57:27.925924234] [21554]  WARN RPiSdn sdn.cpp:40 Using legacy SDN tuning - please consider moving SDN inside rpi.denoise
[1:57:27.933452084] [21554]  INFO RPI vc4.cpp:400 Registered camera /base/soc/i2c0mux/i2c@1/imx708@1a to Unicam device /dev/media2 and ISP device /dev/media5
Using camera /base/soc/i2c0mux/i2c@0/imx708@1a as cam0
[1:57:27.936054990] [21553]  INFO Camera camera.cpp:1183 configuring streams: (0) 800x600-XRGB8888
[1:57:27.937278286] [21554]  INFO RPI vc4.cpp:569 Sensor: /base/soc/i2c0mux/i2c@0/imx708@1a - Selected sensor format: 1536x864-SBGGR10_1X10 - Selected unicam format: 1536x864-pBAA
cam0: Capture 1 frames
7048.315389 (0.00 fps) cam0-stream0 seq: 000007 bytesused: 1920000
```

---

## About the error:

I found out the `isArray_` error comes from building libcamera yourself.
If I use the `ros-humble-libcamera` package it works just fine, no error.

If I compile `camera_ros` with `ros-humble-libcamera` (and then uninstall it and compile `libcamera`) and run it with self-compiled `libcamera` it also shows the error.
I tried this with the latest `libcamera` source, and with `0.1.0`, both behaved the same.

---

> `libcamera` is quite versatile and aims to support all the "standard webcam" hardware. So I think this is a good replacement for the V4L2 and Raspberry Pi specific libraries.

Definitely! There's somewhat of a revolution happening in the linux-camera world.
Pipewire is planning having camera support / already has it.
And V4L2 cameras that needed binary blobs are being replaced with open-source libcamera drivers.

> Can you update the remote branches, hard-reset on the updated remote branch for fix_dynamic_extent, and then try this again?

I did all of this yesterday, the latest commit in `fix_dynamic_extent` was 2 weeks ago
→ nothing changed when I updated just now! :P

I also tried using 6a68576 (checked out to it), it too starts a `1920x1080-NV21` stream

---

> So you are saying that the `Assertion 'isArray_' failed` error that was reported in the other issue does not occur with the binary `ros-humble-libcamera` (version 0.1), but only happens when building versions 0.1 and 0.2 from source?

Exactly. this issue is persistent across all libcamera versions, not only `0.1.0` and `0.2.0`.
Every from-source build causes the error.

I just tried it on jazzy with the testing repo and `ros-jazzy-libcamera`
it compiles and runs just fine, running on `libcamera` `0.3.0`
It also does a `1920x1080-NV21` stream per default :/ 
```console
root@bento-dev:/camera_ws# ros2 run camera_ros camera_node 
[6:00:30.598025078] [4625]  INFO Camera camera_manager.cpp:313 libcamera v0.3.0
[6:00:30.748962703] [4640]  WARN RPiSdn sdn.cpp:40 Using legacy SDN tuning - please consider moving SDN inside rpi.denoise
[6:00:30.751851532] [4640]  INFO RPI vc4.cpp:400 Registered camera /base/soc/i2c0mux/i2c@0/imx708@1a to Unicam device /dev/media4 and ISP device /dev/media1
[6:00:30.752833092] [4640]  WARN CameraSensorProperties camera_sensor_properties.cpp:274 No static properties available for 'imx708_wide'
[6:00:30.752885573] [4640]  WARN CameraSensorProperties camera_sensor_properties.cpp:276 Please consider updating the camera sensor properties database
[6:00:30.762190940] [4640]  WARN RPiSdn sdn.cpp:40 Using legacy SDN tuning - please consider moving SDN inside rpi.denoise
[6:00:30.764633572] [4640]  INFO RPI vc4.cpp:400 Registered camera /base/soc/i2c0mux/i2c@1/imx708@1a to Unicam device /dev/media5 and ISP device /dev/media3
[INFO] [1718347963.455339498] [camera]: 
>> cameras:
   0: imx708 (/base/soc/i2c0mux/i2c@0/imx708@1a)
   1: imx708_wide (/base/soc/i2c0mux/i2c@1/imx708@1a)
[WARN] [1718347963.455850528] [camera]: no camera selected, using default: "/base/soc/i2c0mux/i2c@0/imx708@1a"
[WARN] [1718347963.455905805] [camera]: set parameter 'camera' to silent this warning
[INFO] [1718347963.456957235] [camera]: 
>> stream formats:
   - Pixelformat: NV21 (64x64 - 4608x2592)
   - Pixelformat: XBGR8888 (64x64 - 4608x2592)
   - Pixelformat: BGR888 (64x64 - 4608x2592)
   - Pixelformat: RGB888 (64x64 - 4608x2592)
   - Pixelformat: XRGB8888 (64x64 - 4608x2592)
   - Pixelformat: YUYV (64x64 - 4608x2592)
   - Pixelformat: UYVY (64x64 - 4608x2592)
[WARN] [1718347963.457257527] [camera]: no pixel format selected, using default: "NV21"
[WARN] [1718347963.457331896] [camera]: set parameter 'format' to silent this warning
[INFO] [1718347963.457832444] [camera]: 
>> NV21 format sizes:
   - 160x120
   - 240x160
   - 320x240
   - 400x240
   - 480x320
   - 640x360
   - 640x480
   - 720x480
   - 768x480
   - 854x480
   - 720x576
   - 800x600
   - 960x540
   - 1024x576
   - 960x640
   - 1024x600
   - 1024x768
   - 1280x720
   - 1152x864
   - 1280x800
   - 1360x768
   - 1366x768
   - 1440x900
   - 1280x1024
   - 1536x864
   - 1280x1080
   - 1600x900
   - 1400x1050
   - 1680x1050
   - 1600x1200
   - 1920x1080
   - 2048x1080
   - 1920x1200
   - 2160x1080
   - 2048x1152
   - 2560x1080
   - 2048x1536
   - 2560x1440
   - 2560x1600
   - 3840x1080
   - 2960x1440
   - 3440x1440
   - 2560x2048
   - 3200x1800
   - 3840x1600
   - 3200x2048
   - 3200x2400
   - 3840x2160
   - 4096x2160
   - 3840x2400
[WARN] [1718347963.458043997] [camera]: no dimensions selected, auto-selecting: "1920x1080"
[WARN] [1718347963.458091644] [camera]: set parameter 'width' or 'height' to silent this warning
[6:00:30.768202002] [4625]  INFO Camera camera.cpp:1183 configuring streams: (0) 1920x1080-NV21
[6:00:30.769263413] [4640]  INFO RPI vc4.cpp:569 Sensor: /base/soc/i2c0mux/i2c@0/imx708@1a - Selected sensor format: 2304x1296-SBGGR10_1X10 - Selected unicam format: 2304x1296-pBAA
[INFO] [1718347963.461402727] [camera]: camera "/base/soc/i2c0mux/i2c@0/imx708@1a" configured with 1920x1080-NV21 stream
[WARN] [1718347963.464597737] [camera]: control NoiseReductionMode (10002) not handled
[6:00:30.786225227] [4649]  WARN IPARPI ipa_base.cpp:1166 Could not set AF_TRIGGER - no AF algorithm or not Auto
[6:00:30.786499130] [4649]  WARN IPARPI ipa_base.cpp:1149 Could not set AF_PAUSE - no AF algorithm or not Continuous
[INFO] [1718347963.907671899] [camera]: using default calibration URL
[INFO] [1718347963.907864230] [camera]: camera calibration URL: file:///root/.ros/camera_info/imx708__base_soc_i2c0mux_i2c_0_imx708_1a_1920x1080.yaml
[ERROR] [1718347963.908123300] [camera_calibration_parsers]: Unable to open camera calibration file [/root/.ros/camera_info/imx708__base_soc_i2c0mux_i2c_0_imx708_1a_1920x1080.yaml]
[WARN] [1718347963.908307871] [camera]: Camera calibration file /root/.ros/camera_info/imx708__base_soc_i2c0mux_i2c_0_imx708_1a_1920x1080.yaml not found
[6:00:31.323487990] [4649]  WARN IPARPI ipa_base.cpp:1166 Could not set AF_TRIGGER - no AF algorithm or not Auto
[6:00:31.323689728] [4649]  WARN IPARPI ipa_base.cpp:1149 Could not set AF_PAUSE - no AF algorithm or not Continuous
```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -35,7 +35,7 @@ jobs:
           apt install -y apt-rdepends ros-${{ matrix.distribution }}-launch-ros
           python3 -m venv ~/.venvs/dev
           . ~/.venvs/dev/bin/activate
-          pip install colcon-ros lark
+          pip install lark
           pip install colcon-lint
 
       - name: build and test

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -715,9 +715,12 @@ CameraNode::onParameterChange(const std::vector<rclcpp::Parameter> &parameters)
         }
 
         const std::size_t extent = get_extent(id);
-        if ((value.isArray() && (extent > 0)) && value.numElements() != extent) {
+        if (value.isArray() &&
+            (extent != libcamera::dynamic_extent) &&
+            (value.numElements() != extent))
+        {
           result.successful = false;
-          result.reason = parameter.get_name() + ": parameter dimensions mismatch, expected " +
+          result.reason = parameter.get_name() + ": array dimensions mismatch, expected " +
                           std::to_string(extent) + ", got " + std::to_string(value.numElements());
           return result;
         }

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -440,39 +440,32 @@ CameraNode::declareParameters()
     // store control id with name
     parameter_ids[id->name()] = id;
 
-    std::size_t extent;
-    try {
-      extent = get_extent(id);
-    }
-    catch (const std::runtime_error &e) {
-      // ignore
-      RCLCPP_WARN_STREAM(get_logger(), e.what());
+    if (info.min().numElements() != info.max().numElements())
+      throw std::runtime_error("minimum and maximum parameter array sizes do not match");
+
+    // check if the control can be mapped to a parameter
+    const rclcpp::ParameterType pv_type = cv_to_pv_type(id);
+    if (pv_type == rclcpp::ParameterType::PARAMETER_NOT_SET) {
+      RCLCPP_WARN_STREAM(get_logger(), "unsupported control '" << id->name() << "'");
       continue;
     }
 
     // format type description
-    const bool scalar = (extent == 0);
-    const bool dynamic = (extent == libcamera::dynamic_extent);
-    const std::string cv_type_descr =
-      scalar ? "scalar" : "array[" + (dynamic ? std::string() : std::to_string(extent)) + "]";
-    const std::string cv_descr =
-      std::to_string(id->type()) + " " + cv_type_descr + " range {" + info.min().toString() +
-      "}..{" + info.max().toString() + "}" +
-      (info.def().isNone() ? std::string {} : " (default: {" + info.def().toString() + "})");
-
-    if (info.min().numElements() != info.max().numElements())
-      throw std::runtime_error("minimum and maximum parameter array sizes do not match");
-
-    // clamp default ControlValue to min/max range and cast ParameterValue
-    rclcpp::ParameterValue value;
+    rcl_interfaces::msg::ParameterDescriptor param_descr;
     try {
-      value = cv_to_pv(clamp(info.def(), info.min(), info.max()));
+      const std::size_t extent = get_extent(id);
+      const bool scalar = (extent == 0);
+      const bool dynamic = (extent == libcamera::dynamic_extent);
+      const std::string cv_type_descr =
+        scalar ? "scalar" : "array[" + (dynamic ? std::string() : std::to_string(extent)) + "]";
+      param_descr.description =
+        std::to_string(id->type()) + " " + cv_type_descr + " range {" + info.min().toString() +
+        "}..{" + info.max().toString() + "}" +
+        (info.def().isNone() ? std::string {} : " (default: {" + info.def().toString() + "})");
     }
-    catch (const invalid_conversion &e) {
-      RCLCPP_ERROR_STREAM(get_logger(), "unsupported control '"
-                                          << id->name()
-                                          << "' (type: " << std::to_string(info.def().type()) << "): "
-                                          << e.what());
+    catch (const std::runtime_error &e) {
+      // ignore
+      RCLCPP_WARN_STREAM(get_logger(), e.what());
       continue;
     }
 
@@ -497,18 +490,30 @@ CameraNode::declareParameters()
       break;
     }
 
-    rcl_interfaces::msg::ParameterDescriptor param_descr;
-    param_descr.description = cv_descr;
     if (range_int.from_value != range_int.to_value)
       param_descr.integer_range = {range_int};
     if (range_float.from_value != range_float.to_value)
       param_descr.floating_point_range = {range_float};
 
+    // clamp default ControlValue to min/max range and cast ParameterValue
+    rclcpp::ParameterValue value;
+    try {
+      value = cv_to_pv(clamp(info.def(), info.min(), info.max()));
+    }
+    catch (const invalid_conversion &e) {
+      RCLCPP_ERROR_STREAM(get_logger(), "unsupported control '"
+                                          << id->name()
+                                          << "' (type: " << std::to_string(info.def().type()) << "): "
+                                          << e.what());
+      continue;
+    }
+
     // declare parameters and set default or initial value
     RCLCPP_DEBUG_STREAM(get_logger(),
                         "declare " << id->name() << " with default " << rclcpp::to_string(value));
+
     if (value.get_type() == rclcpp::ParameterType::PARAMETER_NOT_SET) {
-      declare_parameter(id->name(), cv_to_pv_type(id), param_descr);
+      declare_parameter(id->name(), pv_type, param_descr);
     }
     else {
       declare_parameter(id->name(), value, param_descr);

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -464,7 +464,17 @@ CameraNode::declareParameters()
       throw std::runtime_error("minimum and maximum parameter array sizes do not match");
 
     // clamp default ControlValue to min/max range and cast ParameterValue
-    const rclcpp::ParameterValue value = cv_to_pv(clamp(info.def(), info.min(), info.max()));
+    rclcpp::ParameterValue value;
+    try {
+      value = cv_to_pv(clamp(info.def(), info.min(), info.max()));
+    }
+    catch (const invalid_conversion &e) {
+      RCLCPP_ERROR_STREAM(get_logger(), "unsupported control '"
+                                          << id->name()
+                                          << "' (type: " << std::to_string(info.def().type()) << "): "
+                                          << e.what());
+      continue;
+    }
 
     // get smallest bounds for minimum and maximum set
     rcl_interfaces::msg::IntegerRange range_int;

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -508,7 +508,7 @@ CameraNode::declareParameters()
     RCLCPP_DEBUG_STREAM(get_logger(),
                         "declare " << id->name() << " with default " << rclcpp::to_string(value));
     if (value.get_type() == rclcpp::ParameterType::PARAMETER_NOT_SET) {
-      declare_parameter(id->name(), cv_to_pv_type(id->type(), extent > 0), param_descr);
+      declare_parameter(id->name(), cv_to_pv_type(id), param_descr);
     }
     else {
       declare_parameter(id->name(), value, param_descr);

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -461,7 +461,7 @@ CameraNode::declareParameters()
       throw std::runtime_error("minimum and maximum parameter array sizes do not match");
 
     // clamp default ControlValue to min/max range and cast ParameterValue
-    const rclcpp::ParameterValue value = cv_to_pv(clamp(info.def(), info.min(), info.max()), extent);
+    const rclcpp::ParameterValue value = cv_to_pv(clamp(info.def(), info.min(), info.max()));
 
     // get smallest bounds for minimum and maximum set
     rcl_interfaces::msg::IntegerRange range_int;

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -451,10 +451,13 @@ CameraNode::declareParameters()
     }
 
     // format type description
+    const bool scalar = (extent == 0) || (extent == 1);
+    const bool dynamic = (extent == libcamera::dynamic_extent);
+    const std::string cv_type_descr =
+      scalar ? "scalar" : "array[" + (dynamic ? std::string() : std::to_string(extent)) + "]";
     const std::string cv_descr =
-      std::to_string(id->type()) + " " +
-      std::string(extent > 1 ? "array[" + std::to_string(extent) + "]" : "scalar") + " range {" +
-      info.min().toString() + "}..{" + info.max().toString() + "}" +
+      std::to_string(id->type()) + " " + cv_type_descr + " range {" + info.min().toString() +
+      "}..{" + info.max().toString() + "}" +
       (info.def().isNone() ? std::string {} : " (default: {" + info.def().toString() + "})");
 
     if (info.min().numElements() != info.max().numElements())

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -451,7 +451,7 @@ CameraNode::declareParameters()
     }
 
     // format type description
-    const bool scalar = (extent == 0) || (extent == 1);
+    const bool scalar = (extent == 0);
     const bool dynamic = (extent == libcamera::dynamic_extent);
     const std::string cv_type_descr =
       scalar ? "scalar" : "array[" + (dynamic ? std::string() : std::to_string(extent)) + "]";

--- a/src/clamp.cpp
+++ b/src/clamp.cpp
@@ -181,7 +181,17 @@ less(const libcamera::ControlValue &lhs, const libcamera::ControlValue &rhs)
       return false;
     }
   }
+  else if (rhs.isArray()) {
+    // scalar-array comparison
+    const libcamera::Span<const T> vb = rhs.get<libcamera::Span<const T>>();
+    const T va = lhs.get<T>();
+    for (size_t i = 0; i < vb.size(); i++)
+      if (va < vb[i])
+        return true;
+    return false;
+  }
   else {
+    assert(!lhs.isArray() && !rhs.isArray());
     // scalar-scalar comparison
     return lhs.get<T>() < rhs.get<T>();
   }
@@ -211,7 +221,17 @@ greater(const libcamera::ControlValue &lhs, const libcamera::ControlValue &rhs)
       return false;
     }
   }
+  else if (rhs.isArray()) {
+    // scalar-array comparison
+    const libcamera::Span<const T> vb = rhs.get<libcamera::Span<const T>>();
+    const T va = lhs.get<T>();
+    for (size_t i = 0; i < vb.size(); i++)
+      if (va > vb[i])
+        return true;
+    return false;
+  }
   else {
+    assert(!lhs.isArray() && !rhs.isArray());
     // scalar-scalar comparison
     return lhs.get<T>() > rhs.get<T>();
   }

--- a/src/cv_to_pv.cpp
+++ b/src/cv_to_pv.cpp
@@ -13,7 +13,7 @@
 
 #define CASE_CONVERT(T)           \
   case libcamera::ControlType##T: \
-    return cv_to_pv(extract_value<ControlTypeMap<libcamera::ControlType##T>::type>(value), extent);
+    return cv_to_pv(extract_value<ControlTypeMap<libcamera::ControlType##T>::type>(value));
 
 #define CASE_NONE(T)              \
   case libcamera::ControlType##T: \
@@ -75,25 +75,23 @@ cv_to_pv_scalar(const libcamera::Size &size)
 
 template<typename T>
 rclcpp::ParameterValue
-cv_to_pv(const std::vector<T> &values, const std::size_t &extent)
+cv_to_pv(const std::vector<T> &values)
 {
-  if ((values.size() > 1 && extent > 1) && (values.size() != extent))
-    throw std::runtime_error("type extent (" + std::to_string(extent) + ") and value size (" +
-                             std::to_string(values.size()) + ") cannot be larger than 1 and differ");
-
-  if (values.size() > 1)
-    return cv_to_pv_array(values);
-  else if (values.size() == 1)
-    if (!extent)
-      return cv_to_pv_scalar(values[0]);
-    else
-      return cv_to_pv_array(std::vector<T>(extent, values[0]));
-  else
+  switch (values.size()) {
+  case 0:
+    // empty array
     return rclcpp::ParameterValue();
+  case 1:
+    // single element (scalar)
+    return cv_to_pv_scalar(values[0]);
+  default:
+    // dynamic array
+    return cv_to_pv_array(values);
+  }
 }
 
 rclcpp::ParameterValue
-cv_to_pv(const libcamera::ControlValue &value, const std::size_t &extent)
+cv_to_pv(const libcamera::ControlValue &value)
 {
   switch (value.type()) {
     CASE_NONE(None)

--- a/src/cv_to_pv.cpp
+++ b/src/cv_to_pv.cpp
@@ -48,7 +48,7 @@ template<typename T,
 rclcpp::ParameterValue
 cv_to_pv_array(const std::vector<T> & /*values*/)
 {
-  throw std::runtime_error("ParameterValue only supported for arithmetic types");
+  throw invalid_conversion("ParameterValue only supported for arithmetic types");
 }
 
 template<

--- a/src/cv_to_pv.cpp
+++ b/src/cv_to_pv.cpp
@@ -1,4 +1,5 @@
 #include "cv_to_pv.hpp"
+#include "type_extent.hpp"
 #include "types.hpp"
 #include <cstdint>
 #include <libcamera/base/span.h>
@@ -109,10 +110,10 @@ cv_to_pv(const libcamera::ControlValue &value)
 }
 
 rclcpp::ParameterType
-cv_to_pv_type(const libcamera::ControlType &type, const bool is_array)
+cv_to_pv_type(const libcamera::ControlId *const id)
 {
-  if (!is_array) {
-    switch (type) {
+  if (get_extent(id) == 0) {
+    switch (id->type()) {
     case libcamera::ControlType::ControlTypeNone:
       return rclcpp::ParameterType::PARAMETER_NOT_SET;
     case libcamera::ControlType::ControlTypeBool:
@@ -132,7 +133,7 @@ cv_to_pv_type(const libcamera::ControlType &type, const bool is_array)
     }
   }
   else {
-    switch (type) {
+    switch (id->type()) {
     case libcamera::ControlType::ControlTypeNone:
       return rclcpp::ParameterType::PARAMETER_NOT_SET;
     case libcamera::ControlType::ControlTypeBool:

--- a/src/cv_to_pv.hpp
+++ b/src/cv_to_pv.hpp
@@ -5,7 +5,7 @@
 
 
 rclcpp::ParameterValue
-cv_to_pv(const libcamera::ControlValue &value, const std::size_t &extent);
+cv_to_pv(const libcamera::ControlValue &value);
 
 rclcpp::ParameterType
 cv_to_pv_type(const libcamera::ControlType &type, const bool is_array);

--- a/src/cv_to_pv.hpp
+++ b/src/cv_to_pv.hpp
@@ -4,6 +4,13 @@
 #include <rclcpp/parameter_value.hpp>
 
 
+class invalid_conversion : public std::runtime_error
+{
+public:
+  explicit invalid_conversion(const std::string &msg) : std::runtime_error(msg) {};
+};
+
+
 rclcpp::ParameterValue
 cv_to_pv(const libcamera::ControlValue &value);
 

--- a/src/cv_to_pv.hpp
+++ b/src/cv_to_pv.hpp
@@ -7,7 +7,7 @@
 class invalid_conversion : public std::runtime_error
 {
 public:
-  explicit invalid_conversion(const std::string &msg) : std::runtime_error(msg) {};
+  explicit invalid_conversion(const std::string &msg) : std::runtime_error(msg) {}
 };
 
 

--- a/src/cv_to_pv.hpp
+++ b/src/cv_to_pv.hpp
@@ -15,4 +15,4 @@ rclcpp::ParameterValue
 cv_to_pv(const libcamera::ControlValue &value);
 
 rclcpp::ParameterType
-cv_to_pv_type(const libcamera::ControlType &type, const bool is_array);
+cv_to_pv_type(const libcamera::ControlId *const id);

--- a/src/type_extent.cpp
+++ b/src/type_extent.cpp
@@ -19,6 +19,7 @@ template<typename T, std::enable_if_t<!libcamera::details::is_span<T>::value, bo
 std::size_t
 get_extent(const libcamera::Control<T> &)
 {
+  // return an extent of 0 for non-span types
   return 0;
 }
 
@@ -26,7 +27,12 @@ template<typename T, std::enable_if_t<libcamera::details::is_span<T>::value, boo
 std::size_t
 get_extent(const libcamera::Control<T> &)
 {
-  return libcamera::Control<T>::type::extent;
+  // return the span extent, excluding 0
+  // This assumes that libcamera does not define control types
+  // with a fixed size span that does not hold any elements
+  constexpr std::size_t extent = libcamera::Control<T>::type::extent;
+  static_assert(extent != 0);
+  return extent;
 }
 
 #define IF(T)                                  \
@@ -35,7 +41,7 @@ get_extent(const libcamera::Control<T> &)
 
 
 std::size_t
-get_extent(const libcamera::ControlId *id)
+get_extent(const libcamera::ControlId *const id)
 {
 #if LIBCAMERA_VER_GE(0, 1, 0)
   IF(AeEnable)

--- a/src/type_extent.hpp
+++ b/src/type_extent.hpp
@@ -6,5 +6,11 @@ namespace libcamera
 class ControlId;
 }
 
+/**
+ * @brief get the extent of a libcamera::ControlId
+ * @param id
+ * @return the extent of the control: 0 if the control is not a span,
+ *         otherwise [1 ... libcamera::dynamic_extent]
+ */
 std::size_t
-get_extent(const libcamera::ControlId *id);
+get_extent(const libcamera::ControlId *const id);

